### PR TITLE
Refactor OneDriveApi implementation

### PR DESCRIPTION
**Pull request by JC-comp**
This pull request resolves the OneDriveAPI refactoring part of #2530.

Changes
- Add error handling wrapper for all request to handle retry automatically
- Record the request and response for all requests to facilitate dumping information when errors occur, eliminating the need to reproduce them again with a higher log level
- Replace the use of force exit with throwing errors, allowing the caller to handle them to prevent process loss
- To prevent out-of-order execution caused by acquiring the auth token, all connection information, request headers, and callbacks of the Curl socket have been delayed and set altogether

Notes
- This PR depends on and combined #2607 

What is still missing?
- Refactor callers in Sync.d is pending until the merging and settling of all related tool refactorings
- Replace the force exit operation in the authentication flow
- Beautify logging format

### CODE REVIEW

File: src/util.d [hunk 1]
```diff
@@ -627,7 +627,7 @@ void displayFileSystemErrorMessage(string message, string callingFunction) {
     addLogEntry("  Error Message:    " ~ errorMessage);
     
     // Log the calling function
-    addLogEntry("  Calling Function: " ~ callingFunction, ["verbose"]);
+    addLogEntry("  Calling Function: " ~ callingFunction);
```
Comment (abraunegg) [inline src/util.d:630]: This needs to be only displayed when using --verbose .. not all the time to reduce the error output. Please revert this change

Comment (JC-comp) [reply src/util.d:630]: My idea was that all error-related information valuable for traceability, such as connection details, failed reasons, and exact functions, should be recursively dumped upon encountering an error. This way, we wouldn't need to search and match in the log or attempt to reproduce something when errors occur to obtain a detailed report. Additionally, since this function is only called upon encountering an error, its impact would be minimal when everything just works fine.
Feel free to let me know if you find this approach unnecessary, as it might be influenced by the ongoing PR that will change the behavior of error handling. I'll adjust accordingly.

Comment (abraunegg) [reply src/util.d:630]: The original intention behind all of the error logging was so that, when users (usually 99% of them) are running the application, they are running in a non-verbose mode due to all the Internet help guides out there.

In the early days, a number of users wanted minimal output, minimal repetitive output as well - and when error logging was introduced, it was a good way to separate out the level of error logging output between non-verbose application error logging and what occurs during verbose error logging.

I am OK with you leaving this way it is, but please in future, check with me *why* something is a certain way.

Do to this - please contact me offline via email.

**Comment by abraunegg** [comment]
@JC-comp 
This PR fails to compile:
```
ldc2  -w -g -O -J. -d-debug -gc -L-lcurl -L-lsqlite3  -L-ldl src/main.d src/config.d src/log.d src/util.d src/qxor.d src/curlEngine.d src/onedrive.d src/webhook.d src/sync.d src/itemdb.d src/sqlite.d src/clientSideFiltering.d src/monitor.d src/arsd/cgi.d -ofonedrive
src/arsd/cgi.d(3921): Error: static assert:  "Not implemented in your compiler version!"
src/webhook.d(112):        instantiated from here: serveEmbeddedHttp!(handle, Cgi, 5000000L)
```

So in your testing and building, you *must* test each PR separately, by itself, compiling against the minimum LDC version, using a script similar to the following (assuming you already know how to install 'ldc' manually, and can install v1.20.1):

```
#!/bin/bash
  
PR=2628

rm -rf ./onedrive-pr${PR}
git clone https://github.com/abraunegg/onedrive.git onedrive-pr${PR}
cd onedrive-pr${PR}
git fetch origin pull/${PR}/head:pr${PR}
git checkout pr${PR}

# MIN LDC Version to compile
# MIN Version for ARM / Compiling with LDC
#source ~/dlang/ldc-1.18.0/activate
source ~/dlang/ldc-1.20.1/activate


./configure --enable-debug --enable-notifications; make clean; make;
deactivate
./onedrive --version
```
Post building, then start running all your tests for which the PR is resolving and/or fixing.

**Comment by JC-comp** [comment]
Update
- Merge changes to ensure backward compatibility with ldc v1.20.1.

**Comment by abraunegg** [comment]
@JC-comp 
With this PR, please can you:

* Fix the current conflicts
* Baseline re-merge the current 'alpha-5' into this PR, so that this can be tested as a complete item

**Comment by JC-comp** [comment]
@abraunegg 
'alpha-5' re-based

**Comment by abraunegg** [comment]
@JC-comp 
Please re-fix merge issues

Additionally, this PR breaks the entire fetching the correct re-try value and sets it to a hard coded 300 second value. Please restore original functionality.

**Comment by JC-comp** [comment]
@abraunegg 
Conflicts fixed.

For the hard-coded retry value, it's a redundant function slated for removal in the future sync.d refactoring work. It will never be called in this version, as all retry mechanisms are handled at the OneDrive API level.
https://github.com/abraunegg/onedrive/blob/c8fc1e6963e928497897c08cb9d7595c27a16151/src/onedrive.d#L1311-L1318

**Comment by abraunegg** [comment]
@JC-comp 
For this PR to proceed any further, the following needs to be fixed:
1. Coding style - function declarations are not consistent with the coding style
2. Language in changed logging output needs to be re-instated
3. Output logging as per original functionality needs to be re-instated
4. Example of how this updated code should be used to change all the error handling in sync.d and where required
5. All error handling needs to be handed as per before - there are highly specific parts in the code where by the response of X does Y but in other parts does Z for the same response from the API ... 


Overall the code 'works' but this is IMHO only partially complete. I am not going to consider this for v2.5.0 at this stage as this needs a lot more work.

**Comment by JC-comp** [comment]
@abraunegg 
I'm uncertain what I can do to push this PR further.

1-3 involve coding style and output formatting preferences, which lack formal documentation for strict adherence.

4, for the sync.d refactoring, the essential change introduced by this PR is the removal of the retry mechanism, which is already addressed at the API level, and ensuring appropriate actions are taken when the retry limit is reached.

5 seems ambiguous to me. Aside from implementing a back-off mechanism at higher level, all errors are propagated back to the original calling function. If certain responses trigger different actions, it might suggest a flaw in the design of the sync.d functions.

As this is initial step toward addressing potential bugs and unhandled error cases in `sync.d`, without this PR, I cannot proceed with the subsequent tasks.

**Comment by abraunegg** [comment]
Closing in favour of #2686

**Comment by abraunegg** [comment]
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
